src/main/kotlin/fi/metatavu/jsp/api/OrdersApiImpl.kt

```diff
@@ -39,10 +46,48 @@ class OrdersApiImpl: OrdersApi, AbstractApi() {
                 order.exceptionsFromPlans,
                 loggerUserId!!
         )
+
+        createGenericProducts(order.sinks, createdOrder, GenericProductType.SINK)
+                ?: return createBadRequest("Sinks list can only contain products of type SINK")
+
+        createGenericProducts(order.otherProducts, createdOrder, GenericProductType.OTHER)
+                ?: return createBadRequest("Other products list can only contain products of type OTHER")
+
+        createGenericProducts(order.domesticAppliances, createdOrder, GenericProductType.DOMESTIC_APPLIANCE)
+                ?: return createBadRequest("Domestic appliances list can only contain products of type DOMESTIC_APPLIANCE")
+
+        createGenericProducts(order.electricProducts, createdOrder, GenericProductType.ELECTRIC)
+                ?: return createBadRequest("Electric products list can only contain products of type ELECTRIC")
+
+        createGenericProducts(order.intermediateSpaces, createdOrder, GenericProductType.INTERMEDIATE_SPACE)
+                ?: return createBadRequest("Intermediate spaces list can only contain products of type INTERMEDIATE_SPACE")
+
         val translatedOrder = orderTranslator.translate(createdOrder)
         return createOk(translatedOrder)
     }
 
+    /**
+     * Checks that products match a required type and saves generic products to a database from a list
+     *
+     * @param products products to save
+     * @param order the order that these products are related to
+     * @param requiredType required type
+     *
+     * @return empty object if all type checks are successful and null if any of them fails
+     */
+    private fun createGenericProducts (products: List<GenericProduct>, order: CustomerOrder, requiredType: GenericProductType): Any? {
+        for (product in products) {
+            val type = product.type
+            if (type != requiredType) {
+                return null
+            }
+
+            genericProductsController.create(product.name, type, product.supplier, order, loggerUserId!!)
+        }
+
+        return {}
+    }
```

Comment:
Saving should be separated to its own loop that runs after the type check. Return type Boolean? makes more sense.